README.md
# mcstatus

[![Build Status](https://travis-ci.org/Behoston/mcstatus.svg?branch=master)](https://travis-ci.org/Behoston/mcstatus)
[![codecov](https://codecov.io/gh/Behoston/mcstatus/branch/master/graph/badge.svg)](https://codecov.io/gh/Behoston/mcstatus)

`mcstatus` provides an easy way to query Minecraft servers for any information they can expose.
It provides three modes of access (`query`, `status` and `ping`), the differences of which are listed below in usage.

## Usage

```python
from mcstatus import MinecraftServer

# If you know the host and port, you may skip this and use MinecraftServer("example.org", 1234)
server = MinecraftServer.lookup("example.org:1234")

# 'status' is supported by all Minecraft servers that are version 1.7 or higher.
status = server.status()
print("The server has {0} players and replied in {1} ms".format(status.players.online, status.latency))

# 'ping' is supported by all Minecraft servers that are version 1.7 or higher.
# It is included in a 'status' call, but is exposed separate if you do not require the additional info.
latency = server.ping()
print("The server replied in {0} ms".format(latency))

# 'query' has to be enabled in a servers' server.properties file.
# It may give more information than a ping, such as a full player list or mod information.
query = server.query()
print("The server has the following players online: {0}".format(", ".join(query.players.names)))
```

Command Line Interface
```
$ mcstatus
Usage: mcstatus [OPTIONS] ADDRESS COMMAND [ARGS]...

  mcstatus provides an easy way to query Minecraft servers for any
  information they can expose. It provides three modes of access: query,
  status, and ping.

  Examples:

  $ mcstatus example.org ping
  21.120ms

  $ mcstatus example.org:1234 ping
  159.903ms

  $ mcstatus example.org status
  version: v1.8.8 (protocol 47)
  description: "A Minecraft Server"
  players: 1/20 ['Dinnerbone (61699b2e-d327-4a01-9f1e-0ea8c3f06bc6)']

  $ mcstatus example.org query
  host: 93.148.216.34:25565
  software: v1.8.8 vanilla
  plugins: []
  motd: "A Minecraft Server"
  players: 1/20 ['Dinnerbone (61699b2e-d327-4a01-9f1e-0ea8c3f06bc6)']

Options:
  -h, --help  Show this message and exit.

Commands:
  ping    prints server latency
  query   detailed server information
  status  basic server information
```

## Installation

mcstatus is available on pypi, and can be installed trivially with:

```bash
pip install mcstatus
```

Alternatively, just clone this repo!

## License

mcstatus is licensed under Apache 2.0.


## Dev

### Package management

Always work in virtualenv!

```bash
pip install pip-tools
```

Then use [scripts](./scripts)

### Test

```bash
nosetests --with-coverage
```

### Static analysis
```bash
flake8
```

### Plans

- unified API


CHANGELOG
3.0.0
    * DROP Python2 support! Now we can develop at full speed!

2.2
    * Added new CLI
    * Clean up sockets better
    * Fixed connecting to ports greater than 32767
    * Took a very very long time. :D

2.1
    * Fixed the connections hanging when no data is received
    * Sped up Query requests by caching DNS
    * Fixed running setup.py when Six is not installed

2.0
    * Rewrote all the things!
    * Removed CLI. May add it back in future, if there's demand.
    * Supports query and status.
    * Supports SRV lookups.
    * Many tests! Python 3 support! Extra hugs!

1.0

    * Initial release.


mcstatus/tests/test_pinger.py
from unittest import TestCase

from mcstatus.pinger import ServerPinger, PingResponse
from mcstatus.protocol.connection import Connection


class TestServerPinger(TestCase):
    def setUp(self):
        self.pinger = ServerPinger(Connection(), host="localhost", port=25565, version=44)

    def test_handshake(self):
        self.pinger.handshake()

        self.assertEqual(self.pinger.connection.flush(), bytearray.fromhex("0F002C096C6F63616C686F737463DD01"))

    def test_read_status(self):
        raw_response_hex = ("7200707B226465736372697074696F6E223A2241204D696E65637261667420536572766572222C22706"
                            "C6179657273223A7B226D6178223A32302C226F6E6C696E65223A307D2C2276657273696F6E223A7B22"
                            "6E616D65223A22312E382D70726531222C2270726F746F636F6C223A34347D7D")
        raw_response_dict = {
            "description": "A Minecraft Server",
            "players": {
                "max": 20,
                "online": 0,
            },
            "version": {
                "name": "1.8-pre1",
                "protocol": 44,
            },
        }

        self.pinger.connection.receive(bytearray.fromhex(raw_response_hex))
        status = self.pinger.read_status()

        self.assertEqual(status.raw, raw_response_dict)
        self.assertEqual(self.pinger.connection.flush(), bytearray.fromhex("0100"))

    def test_read_status_invalid_json(self):
        self.pinger.connection.receive(bytearray.fromhex("0300017B"))
        self.assertRaises(IOError, self.pinger.read_status)

    def test_read_status_invalid_reply(self):
        raw_response_hex = ("4F004D7B22706C6179657273223A7B226D6178223A32302C226F6E6C696E65223A307D2C2276657"
                            "273696F6E223A7B226E616D65223A22312E382D70726531222C2270726F746F636F6C223A34347D7D")
        self.pinger.connection.receive(bytearray.fromhex(raw_response_hex))

        self.assertRaises(IOError, self.pinger.read_status)

    def test_read_status_invalid_status(self):
        self.pinger.connection.receive(bytearray.fromhex("0105"))

        self.assertRaises(IOError, self.pinger.read_status)

    def test_test_ping(self):
        self.pinger.connection.receive(bytearray.fromhex("09010000000000DD7D1C"))
        self.pinger.ping_token = 14515484

        self.assertTrue(self.pinger.test_ping() >= 0)
        self.assertEqual(self.pinger.connection.flush(), bytearray.fromhex("09010000000000DD7D1C"))

    def test_test_ping_invalid(self):
        self.pinger.connection.receive(bytearray.fromhex("011F"))
        self.pinger.ping_token = 14515484

        self.assertRaises(IOError, self.pinger.test_ping)

    def test_test_ping_wrong_token(self):
        self.pinger.connection.receive(bytearray.fromhex("09010000000000DD7D1C"))
        self.pinger.ping_token = 12345

        self.assertRaises(IOError, self.pinger.test_ping)


class TestPingResponse(TestCase):
    def test_raw(self):
        raw_response = {
            "description": "A Minecraft Server",
            "players": {
                "max": 20,
                "online": 0,
            },
            "version": {
                "name": "1.8-pre1",
                "protocol": 44,
            },
        }
        response = PingResponse(raw_response)
        self.assertEqual(response.raw, raw_response)

    def test_description(self):
        raw_response = {
            "description": "A Minecraft Server",
            "players": {
                "max": 20,
                "online": 0,
            },
            "version": {
                "name": "1.8-pre1",
                "protocol": 44,
            },
        }
        response = PingResponse(raw_response)
        self.assertEqual(response.description, "A Minecraft Server")

    def test_description_missing(self):
        raw_response_without_description = {
            "players": {
                "max": 20,
                "online": 0,
            },
            "version": {
                "name": "1.8-pre1",
                "protocol": 44,
            },
        }
        self.assertRaises(ValueError, PingResponse, raw_response_without_description)

    def test_version(self):
        raw_response = {
            "description": "A Minecraft Server",
            "players": {
                "max": 20,
                "online": 0,
            },
            "version": {
                "name": "1.8-pre1",
                "protocol": 44,
            },
        }
        response = PingResponse(raw_response)

        self.assertIsNotNone(response.version)
        self.assertEqual(response.version.name, "1.8-pre1")
        self.assertEqual(response.version.protocol, 44)

    def test_version_missing(self):
        raw_response_without_version = {
            "description": "A Minecraft Server",
            "players": {
                "max": 20,
                "online": 0,
            },
        }
        self.assertRaises(ValueError, PingResponse, raw_response_without_version)

    def test_version_invalid(self):
        raw_response_with_invalid_version = {
            "description": "A Minecraft Server",
            "players": {
                "max": 20,
                "online": 0,
            },
            "version": "foo",
        }
        self.assertRaises(ValueError, PingResponse, raw_response_with_invalid_version)

    def test_players(self):
        raw_response = {
            "description": "A Minecraft Server",
            "players": {
                "max": 20,
                "online": 5,
            },
            "version": {
                "name": "1.8-pre1",
                "protocol": 44,
            },
        }
        response = PingResponse(raw_response)

        self.assertIsNotNone(response.players)
        self.assertEqual(response.players.max, 20)
        self.assertEqual(response.players.online, 5)

    def test_players_missing(self):
        raw_response_without_players = {
            "description": "A Minecraft Server",
            "version": {
                "name": "1.8-pre1",
                "protocol": 44,
            },
        }
        self.assertRaises(ValueError, PingResponse, raw_response_without_players)

    def test_favicon(self):
        raw_response_with_favicon = {
            "description": "A Minecraft Server",
            "players": {
                "max": 20,
                "online": 0,
            },
            "version": {
                "name": "1.8-pre1",
                "protocol": 44,
            },
            "favicon": "data:image/png;base64,foo",
        }
        response = PingResponse(raw_response_with_favicon)

        self.assertEqual(response.favicon, "data:image/png;base64,foo")

    def test_favicon_missing(self):
        raw_response_without_favicon = {
            "description": "A Minecraft Server",
            "players": {
                "max": 20,
                "online": 0,
            },
            "version": {
                "name": "1.8-pre1",
                "protocol": 44,
            },
        }
        response = PingResponse(raw_response_without_favicon)

        self.assertIsNone(response.favicon)


class TestPingResponsePlayers(TestCase):
    def test_invalid(self):
        self.assertRaises(ValueError, PingResponse.Players, "foo")

    def test_max_missing(self):
        self.assertRaises(ValueError, PingResponse.Players, {"online": 5})

    def test_max_invalid(self):
        self.assertRaises(ValueError, PingResponse.Players, {"max": "foo", "online": 5})

    def test_online_missing(self):
        self.assertRaises(ValueError, PingResponse.Players, {"max": 20})

    def test_online_invalid(self):
        self.assertRaises(ValueError, PingResponse.Players, {"max": 20, "online": "foo"})

    def test_valid(self):
        players = PingResponse.Players({"max": 20, "online": 5})

        self.assertEqual(players.max, 20)
        self.assertEqual(players.online, 5)

    def test_sample(self):
        players = PingResponse.Players(
            {"max": 20, "online": 1, "sample": [{"name": 'Dinnerbone', 'id': "61699b2e-d327-4a01-9f1e-0ea8c3f06bc6"}]})

        self.assertIsNotNone(players.sample)
        self.assertEqual(players.sample[0].name, "Dinnerbone")

    def test_sample_invalid(self):
        self.assertRaises(ValueError, PingResponse.Players, {"max": 20, "online": 1, "sample": "foo"})

    def test_sample_missing(self):
        players = PingResponse.Players({"max": 20, "online": 1})
        self.assertIsNone(players.sample)


class TestPingResponsePlayersPlayer(TestCase):
    def test_invalid(self):
        self.assertRaises(ValueError, PingResponse.Players.Player, "foo")

    def test_name_missing(self):
        self.assertRaises(ValueError, PingResponse.Players.Player, {"id": "61699b2e-d327-4a01-9f1e-0ea8c3f06bc6"})

    def test_name_invalid(self):
        self.assertRaises(ValueError, PingResponse.Players.Player,
                          {"name": {}, "id": "61699b2e-d327-4a01-9f1e-0ea8c3f06bc6"})

    def test_id_missing(self):
        self.assertRaises(ValueError, PingResponse.Players.Player, {"name": "Dinnerbone"})

    def test_id_invalid(self):
        self.assertRaises(ValueError, PingResponse.Players.Player, {"name": "Dinnerbone", "id": {}})

    def test_valid(self):
        player = PingResponse.Players.Player({"name": 'Dinnerbone', 'id': "61699b2e-d327-4a01-9f1e-0ea8c3f06bc6"})

        self.assertEqual(player.name, "Dinnerbone")
        self.assertEqual(player.id, "61699b2e-d327-4a01-9f1e-0ea8c3f06bc6")


class TestPingResponseVersion(TestCase):
    def test_invalid(self):
        self.assertRaises(ValueError, PingResponse.Version, "foo")

    def test_protocol_missing(self):
        self.assertRaises(ValueError, PingResponse.Version, {"name": "foo"})

    def test_protocol_invalid(self):
        self.assertRaises(ValueError, PingResponse.Version, {"name": "foo", "protocol": "bar"})

    def test_name_missing(self):
        self.assertRaises(ValueError, PingResponse.Version, {"protocol": 5})

    def test_name_invalid(self):
        self.assertRaises(ValueError, PingResponse.Version, {"name": {}, "protocol": 5})

    def test_valid(self):
        players = PingResponse.Version({"name": "foo", "protocol": 5})

        self.assertEqual(players.name, "foo")
        self.assertEqual(players.protocol, 5)


mcstatus/tests/test_server.py
from unittest import TestCase
from unittest.mock import patch, Mock

from mcstatus.protocol.connection import Connection
from mcstatus.server import MinecraftServer


class TestMinecraftServer(TestCase):
    def setUp(self):
        self.socket = Connection()
        self.server = MinecraftServer("localhost", port=25565)

    def test_ping(self):
        self.socket.receive(bytearray.fromhex("09010000000001C54246"))

        with patch("mcstatus.server.TCPSocketConnection") as connection:
            connection.return_value = self.socket
            latency = self.server.ping(ping_token=29704774, version=47)

        self.assertEqual(self.socket.flush(), bytearray.fromhex(
            "0F002F096C6F63616C686F737463DD0109010000000001C54246"))
        self.assertEqual(self.socket.remaining(), 0, msg="Data is pending to be read, but should be empty")
        self.assertTrue(latency >= 0)

    def test_ping_retry(self):
        with patch("mcstatus.server.TCPSocketConnection") as connection:
            connection.return_value = None
            with patch("mcstatus.server.ServerPinger") as pinger:
                pinger.side_effect = [Exception, Exception, Exception]
                self.assertRaises(Exception, self.server.ping)
                self.assertEqual(pinger.call_count, 3)

    def test_status(self):
        self.socket.receive(bytearray.fromhex(
            "6D006B7B226465736372697074696F6E223A2241204D696E65637261667420536572766572222C22706C6179657273223A7B226D6"
            "178223A32302C226F6E6C696E65223A307D2C2276657273696F6E223A7B226E616D65223A22312E38222C2270726F746F636F6C22"
            "3A34377D7D09010000000001C54246"
        ))

        with patch("mcstatus.server.TCPSocketConnection") as connection:
            connection.return_value = self.socket
            info = self.server.status(ping_token=29704774, version=47)

        self.assertEqual(self.socket.flush(),
                         bytearray.fromhex("0F002F096C6F63616C686F737463DD01010009010000000001C54246"))
        self.assertEqual(self.socket.remaining(), 0, msg="Data is pending to be read, but should be empty")
        self.assertEqual(info.raw, {"description": "A Minecraft Server", "players": {"max": 20, "online": 0},
                                    "version": {"name": "1.8", "protocol": 47}})
        self.assertTrue(info.latency >= 0)

    def test_status_retry(self):
        with patch("mcstatus.server.TCPSocketConnection") as connection:
            connection.return_value = None
            with patch("mcstatus.server.ServerPinger") as pinger:
                pinger.side_effect = [Exception, Exception, Exception]
                self.assertRaises(Exception, self.server.status)
                self.assertEqual(pinger.call_count, 3)

    def test_query(self):
        self.socket.receive(bytearray.fromhex("090000000035373033353037373800"))
        self.socket.receive(bytearray.fromhex(
            "00000000000000000000000000000000686f73746e616d650041204d696e656372616674205365727665720067616d6574797065"
            "00534d500067616d655f6964004d494e4543524146540076657273696f6e00312e3800706c7567696e7300006d617000776f726c"
            "64006e756d706c61796572730033006d6178706c617965727300323000686f7374706f727400323535363500686f737469700031"
            "39322e3136382e35362e31000001706c617965725f000044696e6e6572626f6e6500446a696e6e69626f6e650053746576650000"
        ))

        self.socket.remaining = Mock()
        self.socket.remaining.side_effect = [15, 208]

        with patch("mcstatus.server.UDPSocketConnection") as connection:
            connection.return_value = self.socket
            info = self.server.query()

        self.assertEqual(self.socket.flush(), bytearray.fromhex(
            "FEFD090000000000000000FEFD000000000021FEDCBA00000000"))
        self.assertEqual(info.raw, {
            "hostname": "A Minecraft Server",
            "gametype": "SMP",
            "game_id": "MINECRAFT",
            "version": "1.8",
            "plugins": "",
            "map": "world",
            "numplayers": "3",
            "maxplayers": "20",
            "hostport": "25565",
            "hostip": "192.168.56.1",
        })

    def test_query_retry(self):
        with patch("mcstatus.server.UDPSocketConnection") as connection:
            connection.return_value = None
            with patch("mcstatus.server.ServerQuerier") as querier:
                querier.side_effect = [Exception, Exception, Exception]
                self.assertRaises(Exception, self.server.query)
                self.assertEqual(querier.call_count, 3)

    def test_by_address_no_srv(self):
        with patch("dns.resolver.query") as query:
            query.return_value = []
            self.server = MinecraftServer.lookup("example.org")
            query.assert_called_once_with("_minecraft._tcp.example.org", "SRV")
        self.assertEqual(self.server.host, "example.org")
        self.assertEqual(self.server.port, 25565)

    def test_by_address_invalid_srv(self):
        with patch("dns.resolver.query") as query:
            query.side_effect = [Exception]
            self.server = MinecraftServer.lookup("example.org")
            query.assert_called_once_with("_minecraft._tcp.example.org", "SRV")
        self.assertEqual(self.server.host, "example.org")
        self.assertEqual(self.server.port, 25565)

    def test_by_address_with_srv(self):
        with patch("dns.resolver.query") as query:
            answer = Mock()
            answer.target = "different.example.org."
            answer.port = 12345
            query.return_value = [answer]
            self.server = MinecraftServer.lookup("example.org")
            query.assert_called_once_with("_minecraft._tcp.example.org", "SRV")
        self.assertEqual(self.server.host, "different.example.org")
        self.assertEqual(self.server.port, 12345)

    def test_by_address_with_port(self):
        self.server = MinecraftServer.lookup("example.org:12345")
        self.assertEqual(self.server.host, "example.org")
        self.assertEqual(self.server.port, 12345)

    def test_by_address_with_multiple_ports(self):
        self.assertRaises(ValueError, MinecraftServer.lookup, "example.org:12345:6789")

    def test_by_address_with_invalid_port(self):
        self.assertRaises(ValueError, MinecraftServer.lookup, "example.org:port")


mcstatus/tests/test_querier.py
from unittest import TestCase

from mcstatus.protocol.connection import Connection
from mcstatus.querier import ServerQuerier, QueryResponse


class TestQuerier(TestCase):
    def setUp(self):
        self.querier = ServerQuerier(Connection())

    def test_handshake(self):
        self.querier.connection.receive(bytearray.fromhex("090000000035373033353037373800"))
        self.querier.handshake()

        self.assertEqual(self.querier.connection.flush(), bytearray.fromhex("FEFD090000000000000000"))
        self.assertEqual(self.querier.challenge, 570350778)

    def test_query(self):
        self.querier.connection.receive(bytearray.fromhex(
            "00000000000000000000000000000000686f73746e616d650041204d696e656372616674205365727665720067616d65747970650"
            "0534d500067616d655f6964004d494e4543524146540076657273696f6e00312e3800706c7567696e7300006d617000776f726c64"
            "006e756d706c61796572730033006d6178706c617965727300323000686f7374706f727400323535363500686f737469700031393"
            "22e3136382e35362e31000001706c617965725f000044696e6e6572626f6e6500446a696e6e69626f6e650053746576650000"
        ))
        response = self.querier.read_query()

        self.assertEqual(self.querier.connection.flush(), bytearray.fromhex("FEFD00000000000000000000000000"))
        self.assertEqual(response.raw, {
            "hostname": "A Minecraft Server",
            "gametype": "SMP",
            "game_id": "MINECRAFT",
            "version": "1.8",
            "plugins": "",
            "map": "world",
            "numplayers": "3",
            "maxplayers": "20",
            "hostport": "25565",
            "hostip": "192.168.56.1",
        })
        self.assertEqual(response.players.names, ["Dinnerbone", "Djinnibone", "Steve"])


class TestQueryResponse(TestCase):
    def setUp(self):
        self.raw = {
            "hostname": "A Minecraft Server",
            "gametype": "SMP",
            "game_id": "MINECRAFT",
            "version": "1.8",
            "plugins": "",
            "map": "world",
            "numplayers": "3",
            "maxplayers": "20",
            "hostport": "25565",
            "hostip": "192.168.56.1",
        }
        self.players = ["Dinnerbone", "Djinnibone", "Steve"]

    def test_valid(self):
        response = QueryResponse(self.raw, self.players)

        self.assertEqual(response.motd, "A Minecraft Server")
        self.assertEqual(response.map, "world")
        self.assertEqual(response.players.online, 3)
        self.assertEqual(response.players.max, 20)
        self.assertEqual(response.players.names, ["Dinnerbone", "Djinnibone", "Steve"])
        self.assertEqual(response.software.brand, "vanilla")
        self.assertEqual(response.software.version, "1.8")
        self.assertEqual(response.software.plugins, [])


class TestQueryResponsePlayers(TestCase):
    def test_valid(self):
        players = QueryResponse.Players(5, 20, ["Dinnerbone", "Djinnibone", "Steve"])

        self.assertEqual(players.online, 5)
        self.assertEqual(players.max, 20)
        self.assertEqual(players.names, ["Dinnerbone", "Djinnibone", "Steve"])


class TestQueryResponseSoftware(TestCase):
    def test_vanilla(self):
        software = QueryResponse.Software("1.8", "")

        self.assertEqual(software.brand, "vanilla")
        self.assertEqual(software.version, "1.8")
        self.assertEqual(software.plugins, [])

    def test_modded(self):
        software = QueryResponse.Software("1.8", "A modded server: Foo 1.0; Bar 2.0; Baz 3.0")

        self.assertEqual(software.brand, "A modded server")
        self.assertEqual(software.plugins, ["Foo 1.0", "Bar 2.0", "Baz 3.0"])

    def test_modded_no_plugins(self):
        software = QueryResponse.Software("1.8", "A modded server")

        self.assertEqual(software.brand, "A modded server")
        self.assertEqual(software.plugins, [])


mcstatus/pinger.py
import datetime
import json
import random

from mcstatus.protocol.connection import Connection


class ServerPinger:
    def __init__(self, connection, host="", port=0, version=47, ping_token=None):
        if ping_token is None:
            ping_token = random.randint(0, (1 << 63) - 1)
        self.version = version
        self.connection = connection
        self.host = host
        self.port = port
        self.ping_token = ping_token

    def handshake(self):
        packet = Connection()
        packet.write_varint(0)
        packet.write_varint(self.version)
        packet.write_utf(self.host)
        packet.write_ushort(self.port)
        packet.write_varint(1)  # Intention to query status

        self.connection.write_buffer(packet)

    def read_status(self):
        request = Connection()
        request.write_varint(0)  # Request status
        self.connection.write_buffer(request)

        response = self.connection.read_buffer()
        if response.read_varint() != 0:
            raise IOError("Received invalid status response packet.")
        try:
            raw = json.loads(response.read_utf())
        except ValueError:
            raise IOError("Received invalid JSON")
        try:
            return PingResponse(raw)
        except ValueError as e:
            raise IOError("Received invalid status response: {}".format(e))

    def test_ping(self):
        request = Connection()
        request.write_varint(1)  # Test ping
        request.write_long(self.ping_token)
        sent = datetime.datetime.now()
        self.connection.write_buffer(request)

        response = self.connection.read_buffer()
        received = datetime.datetime.now()
        if response.read_varint() != 1:
            raise IOError("Received invalid ping response packet.")
        received_token = response.read_long()
        if received_token != self.ping_token:
            raise IOError(
                "Received mangled ping response packet "
                "(expected token {ping_token}, received {received_token})".format(
                    ping_token=self.ping_token,
                    received_token=received_token,
                ))

        delta = (received - sent)
        # We have no trivial way of getting a time delta :(
        return (delta.days * 24 * 60 * 60 + delta.seconds) * 1000 + delta.microseconds / 1000.0


class PingResponse:
    class Players:
        class Player:
            def __init__(self, raw):
                if type(raw) is not dict:
                    raise ValueError("Invalid player object (expected dict, found {})".format(type(raw)))

                if "name" not in raw:
                    raise ValueError("Invalid player object (no 'name' value)")
                if not isinstance(raw["name"], str):
                    raise ValueError(
                        "Invalid player object (expected 'name' to be str, was {})".format(type(raw["name"])))
                self.name = raw["name"]

                if "id" not in raw:
                    raise ValueError("Invalid player object (no 'id' value)")
                if not isinstance(raw["id"], str):
                    raise ValueError("Invalid player object (expected 'id' to be str, was {})".format(type(raw["id"])))
                self.id = raw["id"]

        def __init__(self, raw):
            if type(raw) is not dict:
                raise ValueError("Invalid players object (expected dict, found {}".format(type(raw)))

            if "online" not in raw:
                raise ValueError("Invalid players object (no 'online' value)")
            if type(raw["online"]) is not int:
                raise ValueError(
                    "Invalid players object (expected 'online' to be int, was {})".format(type(raw["online"])))
            self.online = raw["online"]

            if "max" not in raw:
                raise ValueError("Invalid players object (no 'max' value)")
            if type(raw["max"]) is not int:
                raise ValueError("Invalid players object (expected 'max' to be int, was {})".format(type(raw["max"])))
            self.max = raw["max"]

            if "sample" in raw:
                if type(raw["sample"]) is not list:
                    raise ValueError(
                        "Invalid players object (expected 'sample' to be list, was {})".format(type(raw["max"])))
                self.sample = [PingResponse.Players.Player(p) for p in raw["sample"]]
            else:
                self.sample = None

    class Version:
        def __init__(self, raw):
            if type(raw) is not dict:
                raise ValueError("Invalid version object (expected dict, found {}".format(type(raw)))

            if "name" not in raw:
                raise ValueError("Invalid version object (no 'name' value)")
            if not isinstance(raw["name"], str):
                raise ValueError(
                    "Invalid version object (expected 'name' to be str, was {})".format(type(raw["name"])))
            self.name = raw["name"]

            if "protocol" not in raw:
                raise ValueError("Invalid version object (no 'protocol' value)")
            if type(raw["protocol"]) is not int:
                raise ValueError(
                    "Invalid version object (expected 'protocol' to be int, was {})".format(type(raw["protocol"])))
            self.protocol = raw["protocol"]

    def __init__(self, raw):
        self.raw = raw

        if "players" not in raw:
            raise ValueError("Invalid status object (no 'players' value)")
        self.players = PingResponse.Players(raw["players"])

        if "version" not in raw:
            raise ValueError("Invalid status object (no 'version' value)")
        self.version = PingResponse.Version(raw["version"])

        if "description" not in raw:
            raise ValueError("Invalid status object (no 'description' value)")
        self.description = raw["description"]

        if "favicon" in raw:
            self.favicon = raw["favicon"]
        else:
            self.favicon = None

        self.latency = None


mcstatus/server.py
import dns.resolver

from mcstatus.pinger import ServerPinger
from mcstatus.protocol.connection import TCPSocketConnection, UDPSocketConnection
from mcstatus.querier import ServerQuerier


class MinecraftServer:
    def __init__(self, host, port=25565):
        self.host = host
        self.port = port

    @staticmethod
    def lookup(address):
        host = address
        port = None
        if ":" in address:
            parts = address.split(":")
            if len(parts) > 2:
                raise ValueError(f"Invalid address '{address}'")
            host = parts[0]
            port = int(parts[1])
        if port is None:
            port = 25565
            try:
                answers = dns.resolver.query(f"_minecraft._tcp.{host}", "SRV")
                if len(answers):
                    answer = answers[0]
                    host = str(answer.target).rstrip(".")
                    port = int(answer.port)
            except Exception:
                pass

        return MinecraftServer(host, port)

    def ping(self, retries=3, **kwargs):
        connection = TCPSocketConnection((self.host, self.port))
        exception = None
        for attempt in range(retries):
            try:
                pinger = ServerPinger(connection, host=self.host, port=self.port, **kwargs)
                pinger.handshake()
                return pinger.test_ping()
            except Exception as e:
                exception = e
        else:
            raise exception

    def status(self, retries=3, **kwargs):
        connection = TCPSocketConnection((self.host, self.port))
        exception = None
        for attempt in range(retries):
            try:
                pinger = ServerPinger(connection, host=self.host, port=self.port, **kwargs)
                pinger.handshake()
                result = pinger.read_status()
                result.latency = pinger.test_ping()
                return result
            except Exception as e:
                exception = e
        else:
            raise exception

    def query(self, retries=3):
        exception = None
        host = self.host
        try:
            answers = dns.resolver.query(host, "A")
            if len(answers):
                answer = answers[0]
                host = str(answer).rstrip(".")
        except Exception as e:
            pass
        for attempt in range(retries):
            try:
                connection = UDPSocketConnection((host, self.port))
                querier = ServerQuerier(connection)
                querier.handshake()
                return querier.read_query()
            except Exception as e:
                exception = e
        else:
            raise exception


mcstatus/scripts/mcstatus.py
import click

from mcstatus.server import MinecraftServer

server = None


@click.group(context_settings=dict(help_option_names=['-h', '--help']))
@click.argument("address")
def cli(address):
    """
    mcstatus provides an easy way to query Minecraft servers for
    any information they can expose. It provides three modes of
    access: query, status, and ping.

    Examples:

    \b
    $ mcstatus example.org ping
    21.120ms

    \b
    $ mcstatus example.org:1234 ping
    159.903ms

    \b
    $ mcstatus example.org status
    version: v1.8.8 (protocol 47)
    description: "A Minecraft Server"
    players: 1/20 ['Dinnerbone (61699b2e-d327-4a01-9f1e-0ea8c3f06bc6)']

    \b
    $ mcstatus example.org query
    host: 93.148.216.34:25565
    software: v1.8.8 vanilla
    plugins: []
    motd: "A Minecraft Server"
    players: 1/20 ['Dinnerbone (61699b2e-d327-4a01-9f1e-0ea8c3f06bc6)']
    """
    global server
    server = MinecraftServer.lookup(address)


@cli.command(short_help="prints server latency")
def ping():
    """
    Ping server for latency.
    """
    click.echo("{}ms".format(server.ping()))


@cli.command(short_help="basic server information")
def status():
    """
    Prints server status. Supported by all Minecraft
    servers that are version 1.7 or higher.
    """
    response = server.status()
    click.echo("version: v{} (protocol {})".format(response.version.name, response.version.protocol))
    click.echo("description: '{}'".format(response.description))
    click.echo(
        "players: {players_online}/{players_max} {players_list}".format(
            players_online=response.players.online,
            players_max=response.players.max,
            players_list=[
                "{} ({})".format(player.name, player.id)
                for player in response.players.sample
            ] if response.players.sample is not None else "No players online"
        )
    )


@cli.command(short_help="detailed server information")
def query():
    """
    Prints detailed server information. Must be enabled in
    servers' server.properties file.
    """
    response = server.query()
    click.echo("host: {}:{}".format(response.raw['hostip'], response.raw['hostport']))
    click.echo("software: v{version} {brand}".format(
        version=response.software.version,
        brand=response.software.brand,
    ))
    click.echo("plugins: {}".format(response.software.plugins))
    click.echo("motd: '{}'".format(response.motd))
    click.echo(
        "players: {players_online}/{players_max} {players_list}".format(
            players_online=response.players.online,
            players_max=response.players.max,
            players_list=response.players.names,
        )
    )


if __name__ == "__main__":
    cli()
